### ISSUE TITLE
Engine attempting to access excluded files

### ISSUE THREAD

**Issue by elwayman02** [codeclimate/codeclimate#82]
I'm on codeclimate 0.2.8 running the following config:

```
---
engines:
  eslint:
    enabled: true
  csslint:
    enabled: false
  fixme:
    enabled: true
  watson:
    enabled: true
ratings:
  paths:
  - "app/**"
exclude_paths:
- bower_components/**/*
- dist/**/*
- node_modules/**/*
- tests/**/*
- tmp/**/*
- vendor/**/*
```

I get the following error for `codeclimate analyze`:

```
Starting analysis
Running eslint: Done!
error: (CC::Analyzer::Engine::EngineFailure) engine eslint failed with status 1 and stderr
/usr/src/app/node_modules/glob/sync.js:321
        throw er
              ^
Error: EACCES, permission denied '/code/tmp/asset_rewrite-cache_path-RFcmONBd.tmp'
    at Error (native)
    at Object.fs.readdirSync (fs.js:761:18)
    at GlobSync._readdir (/usr/src/app/node_modules/glob/sync.js:275:41)
    at GlobSync._readdirInGlobStar (/usr/src/app/node_modules/glob/sync.js:254:20)
    at GlobSync._readdir (/usr/src/app/node_modules/glob/sync.js:263:17)
    at GlobSync._processGlobStar (/usr/src/app/node_modules/glob/sync.js:330:22)
    at GlobSync._process (/usr/src/app/node_modules/glob/sync.js:128:10)
    at GlobSync._processGlobStar (/usr/src/app/node_modules/glob/sync.js:363:10)
    at GlobSync._process (/usr/src/app/node_modules/glob/sync.js:128:10)
    at GlobSync._processGlobStar (/usr/src/app/node_modules/glob/sync.js:363:10)
```

The project is an Ember 1.12 app w/ over 32,000 lines of code across 200+ files.

I thought the issue was related to https://github.com/codeclimate/codeclimate-eslint/issues/1, but even disabling eslint surfaces the same error for the fixme engine instead.  Seems to be something with the CLI itself, then.

**Comment by elwayman02** [comment]
Anyone there? :(

**Comment by ABaldwinHunter** [comment]
@elwayman02 Thanks for the feedback and apologies for delay! We're looking into this issue.

I'm surprised the error showed up there, because I see that you've excluded `node_modules`. I'd like to reproduce the issue. Are you able to share repo / is it public?

**Comment by elwayman02** [comment]
The repo is not public, unfortunately.

**Comment by elwayman02** [comment]
There are only 3 directories not excluded by my config: 
```
app/
config/
public/
```

**Comment by ABaldwinHunter** [comment]
@elwayman02 Hm I notice that you've chosen to rate `app/**`, but exclude `node_modules`, which is under app.

Have you tried adding `/app/node_modules/**/*` to `exclude_paths`?

**Comment by ABaldwinHunter** [comment]
(The paths in `exclude_paths` should be relative to root)

**Comment by elwayman02** [comment]
In Ember projects, `app/` is not the root directory.  It is a neighbor to `node_modules`.  The project structure looks like this:

```
app/
adapters/
components/
controllers/
etc...
bower_components/
config/
dist/
node_modules/
public/
tests/
tmp/
vendor/
.codeclimate.yml
```

**Comment by ABaldwinHunter** [comment]
@elwayman02 Ah you're right. That seems fine.

Taking another look at the original error, it looks like this is a permissions issue. Are there any special permissions on that directory?

Related issue: https://github.com/codeclimate/codeclimate-eslint/issues/3

I'm investigating to reproduce locally.

**Comment by elwayman02** [comment]
The default permissions on my project directory are `drwxr-xr-x  29 jordan  staff   986`.  How can I make sure docker has write access?

**Comment by pbrisbin** [comment]
You can check the permissions directly with

```
stat tmp/asset_rewrite-cache_path-RFcmONBd.tmp
```

**Comment by elwayman02** [comment]
Ran it again (so the filename will be different):
```
16777220 35719825 drwx------ 15 jordan staff 0 510 "Aug 11 11:23:28 2015" "Aug  6 19:16:45 2015" "Aug  6 19:16:45 2015" "Aug  6 19:16:33 2015" 4096 0 0 tmp/asset_rewrite-cache_path-YSmsAl3D.tmp/
```

**Comment by elwayman02** [comment]
Any ideas on how I can fix this?

**Comment by ABaldwinHunter** [comment]
@elwayman02 Thanks for bringing this issue to our attention. We realize that it's non-trivial and are investing engineering energy in a solution that creates a better architecture to how engines handle permissions issues.

As a short term fix, you can change permissions to 644 by running:
`chmod 644 dir/foo.bar`

**Comment by elwayman02** [comment]
Ok, thanks!

**Comment by ABaldwinHunter** [comment]
Yep. And `chmod 755 foo` to make a directory `drwxr-xr-x`

**Comment by elwayman02** [comment]
Running those commands didn't help, because the file itself is created as part of the analysis and I can't fix the permissions until after the analysis has failed.  I can fix the directory permissions, but it doesn't stop the analysis from failing.

**Comment by ABaldwinHunter** [comment]
@elwayman02 Thanks for the helpful feedback. This issue seems specific to certain checks in Eslint. Could there be a plugin that you're using that creates those files?

There may be some workarounds with manually setting the permissions on files created in the directory, but we realize it's an issue with our `exclude` architecture.

We're working on a solution for this issue now.

**Comment by elwayman02** [comment]
We don't have any special configs/plugins for ESlint in our project (not even .eslintrc).  I was trying to use the defaults from codeclimate just to get it working first.

**Comment by elwayman02** [comment]
@ABaldwinHunter I ran across another related issue, which I can log separately if you prefer.  Once I ran the chmod 644 on the files in `tmp/`, it broke my cleanup scripts for the project.  I couldn't clean `tmp/`, because rimraf didn't have permission to remove the files created by codeclimate.  I can't run the clean script with sudo, because the script runs some other commands like `npm cache clean` and `bower cache clean` which don't necessarily work with sudo.  Something to keep in mind as you address the permissions issues for engines :)

**Comment by ABaldwinHunter** [comment]
@elwayman02 Thanks for the feedback! I'm sorry my suggestion created a new issue for you. We'll keep that information in mind.

**Comment by pbrisbin** [comment]
@elwayman02 we're still working on a fix for these issues, but we had a few questions and I'm wondering if you'd be willing to help us better understand what's going on:

1. You mentioned a script that cleans the `tmp/` directory of these files that codeclimate can't analyze. If you run this script immediately before analysis, does it still error?

  Note: you may need to do a quick `sudo rm -r tmp && mkdir tmp` to correct the permissions from our earlier attempted fixes before trying this.

2. The main source of trouble is that these files are generated into the source directory itself and therefore show up as potentially analyzable. Do you know if there's a way to tell eslint (perhaps via `.eslintrc`) to put these files in `/tmp` (note the leading slash), not `tmp`?

Also, would you be willing to let us look at the source of the project? If so, you can email me a zip file since you mentioned it's a private repo (pat at codeclimate dot com). That would be the best way to help us sort this out, but I understand if it's not feasible.

**Comment by elwayman02** [comment]
@pbrisbin running my clean script:

`"clean": "npm cache clean && bower cache clean && rm -rf node_modules bower_components dist tmp"`

This does seem to allow the analysis to run successfully!  The problem here is seemingly that codeclimate's "exclude_paths" directive isn't being respected by the engines.  `tmp` should never have been analyzed in the first place, so there shouldn't have been a problem with eslint generating files into that directory.

Unfortunately, I can't share the repo, but this seems to be a breakthrough perhaps?

**Comment by pbrisbin** [comment]
Thanks so much for your help. The root issue is yes, the engine is attempting to read files and/or directories that it can't and shouldn't because they're listed as excluded. We have a good idea of why this is happening and are working on a fix. In the meantime, I think if you make it a habit to clean before running an analysis you should be able to avoid the issue.

Could you confirm one more thing for us? Try doing a clean, then analyze, then analyze again. If the second analyze still succeeds, that'd prove that the analysis itself is not creating these files, it's something else. That wouldn't really change things; we'll still make the same fix and you'll still need the same workaround for now, but it would increase our understanding of the issue and could even point us at a different, easier workaround (for example: maybe your development web server is doing this, and *it* can be easily configured to put these files in `/tmp` instead).

**Comment by elwayman02** [comment]
codeclimate is not creating these files.  The files are created as part of the ember-cli build process.  That said, I don't think it's a reasonable workaround to configure it to write to a different directory.  This is the default setup for all ember-cli apps; unless you convince the ember-cli devs to change how their project is configured, every Ember app that attempts to use codeclimate will run into this issue, unfortunately.

**Comment by pbrisbin** [comment]
Gotcha. Well, thanks for the additional understanding.

**Comment by elwayman02** [comment]
No problem! I'm happy to help provide any info needed to get this resolved, so we can switch to using Engines :)

**Comment by alexpjohnson** [comment]
Any update on this issue?

**Comment by pbrisbin** [comment]
Sorry, there is no update. We're still actively working on a fix for this.

**Comment by pbrisbin** [comment]
All,

We believe we have a fix in the latest version (v0.4.0). Please update to this version and give it a try.

**NOTE** we've made this fix specifically for the eslint engine. If you run into a similar error with another engine, please hang tight. We're working to get all affected engines updated as quickly as we can.

**Comment by elwayman02** [comment]
Thanks @pbrisbin pulling the latest docker image now! :)

**Comment by alexpjohnson** [comment]
ESLint worked for me. Just waiting on Rubocop :) Thanks for all the effort guys.

**Comment by elwayman02** [comment]
Doesn't seem to be working still...I updated codeclimate from docker and pulled the latest eslint engine...the error is different but it's still not actually running the config. The analysis runs for about 5-10 minutes (eslint cli itself takes <30s when run outside of codeclimate), and then I get this:

```
jordan$ codeclimate analyze -f text
Starting analysis
Running eslint: Done!

== Brocfile.js (7 issues) ==
error: (NoMethodError) undefined method `source_buffer_for' for "/code":String
```

This is especially confusing because there is no method (or even text string) `source_buffer_for'` anywhere in my project, which leads me to believe this error is coming from codeclimate.

**Comment by pbrisbin** [comment]
Thanks @elwayman02 that seems very unrelated to this initial issue and I've created #111 to track it.

Given the actual issue here appears to be fixed, I'm going to go ahead and close this.

**Comment by elwayman02** [comment]
Thanks @pbrisbin
